modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/HorizontalCardTag.java

```diff
@@ -69,11 +70,12 @@ public String getTitle() {
 		HorizontalCard horizontalCard = getHorizontalCard();
 
 		if ((_title == null) && (horizontalCard != null)) {
-			title = horizontalCard.getTitle();
+			title = HtmlUtil.escape(horizontalCard.getTitle());
 		}
 
-		return LanguageUtil.get(
-			TagResourceBundleUtil.getResourceBundle(pageContext), title);
+		return HtmlUtil.escape(
```

Comment:
You always want to escape as close to the context as necessary. Escaping here means you get an html escaped title which won't be usable in other contexts. 

This is clear even in this PR where you're passing something that is html-escaped to a JS context and an HTML attribute context, so we're not doing the right escaping.

Since we know React escapes for us, I think the only really necessary place you should escape this might be https://github.com/liferay-frontend/liferay-portal/pull/575/files#diff-a45f3b1f2a6d6fad307b929f02e524e2759fe240dd3bf32d0b73dd5b88035225R193 where you need `HtmlUtil.escapeAttribute` since the input goes into an html attribute.

Reply:
Thanks @jbalsas, I wasn't 100% sure if this was correct so I appreciate the help.

I'm going to try and guess where the changed should be applied, because in my browsers: https://github.com/liferay-frontend/liferay-portal/pull/575/files#diff-a45f3b1f2a6d6fad307b929f02e524e2759fe240dd3bf32d0b73dd5b88035225R193 doesn't display anything useful, since we're talking about attributes I think I get the idea.

Reply:
Hey @julien, you should not have to guess but to educate yourself in this. Take your time to understand the problem and ask when in doubt. Otherwise, this becomes a game of wackamole 😉 

[Cross Site Scripting Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html)
- [RULE #0 - Never Insert Untrusted Data Except in Allowed Locations](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#rule-0-never-insert-untrusted-data-except-in-allowed-locations)
- [RULE #1 - HTML Encode Before Inserting Untrusted Data into HTML Element Content](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#rule-1-html-encode-before-inserting-untrusted-data-into-html-element-content)
- [RULE #2 - Attribute Encode Before Inserting Untrusted Data into HTML Common Attributes](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#rule-2-attribute-encode-before-inserting-untrusted-data-into-html-common-attributes)
- [RULE #3 - JavaScript Encode Before Inserting Untrusted Data into JavaScript Data Values](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#rule-3-javascript-encode-before-inserting-untrusted-data-into-javascript-data-values)

If you check our [`HtmlUtil`](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/util/HtmlUtil.java) class:

>  Provides utility methods for escaping, rendering, replacing, and stripping HTML text. This class uses XSS recommendations from <a href="http://www.owasp.org/index.php/Cross_Site_Scripting#How_to_Protect_Yourself">http://www.owasp.org/index.php/Cross_Site_Scripting#How_to_Protect_Yourself</a> when escaping HTML text

| Method | Description |
| --- | --- |
| [escape](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/util/HtmlUtil.java#L43) | Escapes the text so that it is safe to use in an HTML context |
| [escapeAttribute](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/util/HtmlUtil.java#L73) | Escapes the attribute value so that it is safe to use as an attribute value | 
| [escapeCSS](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/util/HtmlUtil.java#L84) | Escapes the CSS value so that it is safe to use in a CSS context |
| [escapeHREF](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/util/HtmlUtil.java#L84) | Escapes the HREF attribute so that it is safe to use as an HREF attribute |
| [escapeJS](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/util/HtmlUtil.java#L108) | Escapes the JavaScript value so that it is safe to use in a JavaScript context | 
| ... | ... | 

So, in your case, since you're escaping for `<div title="escaped_title">`, you want to use a more specific escaping method.

Reply:
thanks, bookmarked.